File: leetcode-cn/215.数组中的第k个最大元素.py
#
# @lc app=leetcode.cn id=215 lang=python3
#
# [215] 数组中的第K个最大元素
#

# @lc code=start
from typing import List


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        def quick_sort(nums, left, right, k):
            if left >= right:
                return nums[left]

            privot = nums[left]

            i = left
            j = right

            while i < j:
                while i < j and nums[j] >= privot:
                    j -= 1
                nums[i] = nums[j]

                while i < j and nums[i] <= privot:
                    i += 1
                nums[j] = nums[i]

            if i == k:
                return privot
            elif k < i:
                return quick_sort(nums, left, i - 1, k)
            else:
                return quick_sort(nums, i + 1, right, k)

        return quick_sort(nums, 0, len(nums) - 1, len(nums) - k)


# @lc code=end

File: leetcode-cn/C++/215.数组中的第k个最大元素.cpp
/*
 * @lc app=leetcode.cn id=215 lang=cpp
 *
 * [215] 数组中的第K个最大元素
 */

// @lc code=start
#include <stdlib.h>
#include <vector>

using namespace std;

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k, int left, int right)
    {
        int i = rand() % (right - left + 1) + left;
        swap(nums[i], nums[left]);

        int p = left;
        int q = right;
        int privot = nums[left];

        while (p < q) {
            while (p < q && nums[q] >= privot) {
                q--;
            }
            nums[p] = nums[q];

            while (p < q && nums[p] <= privot) {
                p++;
            }
            nums[q] = nums[p];
        }

        if (p == k) {
            return privot;
        } else if (p < k) {
            return findKthLargest(nums, k, p + 1, right);
        } else {
            return findKthLargest(nums, k, left, p - 1);
        }
    }

    int findKthLargest(vector<int>& nums, int k)
    {
        return findKthLargest(nums, nums.size() - k, 0, nums.size() - 1);
    }
};
// @lc code=end
